text/0058-standard-library.md

```diff
@@ -0,0 +1,299 @@
+# Cedar Standard Library
+
+## Related issues and PRs
+
+- Reference Issues:
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-03-12
+
+## Summary
+
+Create a Cedar Standard Library of type declarations that are likely to be
+useful to many Cedar users, and are distributed along with Cedar.
+
+Allow users to "extend" the type declarations in the standard library by adding
+more parent types and/or more attributes.
+
+This RFC includes the concept/infrastructure for the Cedar Standard Library, and
+also concretely proposes one type declaration to be included in it: `OIDC::User`.
+Discussion of other type declarations to add to the Cedar Standard Library is
+left out of scope for this RFC.
+
+## Basic example
+
+```
+type User = __cedar::OIDC::User;
+```
+or
+```
+entity Group { myAttribute: String };
+entity User extends __cedar::OIDC::User in [Group] { emailOptOut: Boolean };
```

Comment:
There is also a technical issue: Suppose I have
```
entity Group { myAttribute: String, sibling: [Group] };
entity HierarchicalGroup extends Group { tag: String };
```
A strict interpretation of `extends` would make `HierarchicalGroup` equivalent to 
```
entity HierarchicalGroup { myAttribute: String, sibling: [Group], tag: String };
```
But users might be expecting that `sibling`'s type is updated with the extended type, i.e.,
```
entity HierarchicalGroup { myAttribute: String, sibling: [HierarchicalGroup], tag: String };
```
Doing so would be important for strict validation, since it does not allow subtyping. With permissive validation, leaving `sibling` as `[Group]` is less problematic because `HierarchicalGroups` are effectively subtypes of `Group`s, and we could introduce subtyping edges explicitly to support that.